### PR TITLE
Run experiment in localhost in the same process

### PR DESCRIPTION
**Pull request by crewtool**
# Description
Currently when we run grid it spawns new process which runs cc_train. This PR changes this behaviour, so in localhost it runs in the same process. It enables easier local debugging session as we can run json specified files.

# Checklist
Answer each question with "yes", and provide an explanation if the answer isn't what is expected.

1. Is this PR focused on a single feature without the possibility of being divided into smaller PRs?
   - Answer: yes
2. Are all variables set with logical defaults that are backward compatible?
   - Answer: yes
3. Can you provide a link to a Neptune experiment verifying the functionality of this PR?
   - Answer: NA
4. Have you successfully executed the linter and tests?
   - Answer: yes
5. Are all functions concise and don't require splitting?
   - Answer: yes
6. Is there documentation for arguments and complex logic?
   - Answer: yes
7. Are there no temporary solutions in the code? Should a task be added to Trello to refine the code?
   - Answer: no
8. Are all functions placed in appropriate files?
   - Answer: yes
9. Is the PR name meaningful and descriptive?
   - Answer: yes
10. Is PR description provided, or unnecessary?
    - Answer: yes
  
Other questions:

11. Have you not made changes to the requirements or anything related to the Singularity image? If you have, did you generate and upload a new image to the clusters?
    - Answer: no
12. Is the change significant enough to notify all individuals working with this repository?
    - Answer: yes
13. Is there a need for a second reviewer?
    - Answer: yes

### CODE REVIEW

File: lizrd/scripts/grid.py [hunk 1]
```diff
@@ -232,13 +234,8 @@
...
         elif runner == MachineBackend.LOCAL:
-            subprocess_args = [
-                get_grid_entrypoint(runner),
-                "python3",
-                "-m",
-                RUNNER,
-                *runner_params,
-            ]
+            args = argparse.Namespace(**grid[0])
+            cc_train_main(None, args=args, run_in_the_same_process=True)
```
Comment (maciek-pioro) [inline lizrd/scripts/grid.py:238]: Nie podoba mi się, że to oznacza coupling grida z konkretnym runnerem, ale nie wiem do końca jak to lepiej zrobić. Może chociaż if-ować po `RUNNER` i w zależności od wyniku albo 1. odpalać cc_train_main, 2. działać tak jak było dotąd.

File: research/conditional/train/cc_train.py [hunk 2]
```diff
@@ -56,8 +52,20 @@ def log_batch(dataset_wrapper):
...
+    run_in_the_same_process: bool = False,
 ):
+    if run_in_the_same_process:  # TODO is it necessary?
+        args = vars(args)
+        parser = argparse.ArgumentParser()
+        introduce_parser_arguments(parser)
+        combined_args = vars(parser.parse_known_args()[0])
+        combined_args.update(args)
+        args = argparse.Namespace(**combined_args)
+
```
[the remark this answers is not in the document]
Comment (maciek-pioro) [reply research/conditional/train/cc_train.py:68]: Dodatkowo, trochę bym się bał o typy - jeśli przekazujemy argumenty "przez basha", to nie ma znaczenia czy argument w configu to int czy string. Czy tutaj jest tak samo?

File: lizrd/scripts/grid.py (hunk 1 above)
Comment (crewtool) [reply lizrd/scripts/grid.py]: on-line ustaliliśmy z @maciek-pioro, że w przypadku MachineBackend.LOCAL powinien on najpierw wybrać odpowienią funkcję traina i ją uruchomić.

File: research/conditional/train/cc_train.py [hunk 3]
```diff
@@ -56,8 +52,20 @@ def log_batch(dataset_wrapper):
...
+    rank: Optional[int],
+    data_seeds: Optional[list[int]] = None,
+    port: str = "29500",
+    args: Optional[argparse.Namespace] = None,
+    runner_params: Optional[list] = None,
 ):
+    if runner_params is not None:
+        parser = argparse.ArgumentParser()
+        introduce_parser_arguments(parser)
+        runner_params = process_params(runner_params)
```
Comment (maciek-pioro) [inline research/conditional/train/cc_train.py:64]: Dlaczego tego potrzebnujemy?

**Review by maciek-pioro — APPROVED** [review]

**Comment by maciek-pioro** [comment]
Waiting for reviewer 2

**Review by Simontwice — APPROVED** [review]
lgtm but not an expert

**Pull request merged**